libbpf-tools/javagc.c

```diff
@@ -0,0 +1,243 @@
+			err = errno;
+			fprintf(stderr, "invalid time: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return err;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && ! env.verbose)
+		return 0;
+
+	return vfprintf(stderr, format, args);
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct data_t *e = (struct data_t *)data;
+	struct tm *tm = NULL;
+	char ts[16];
+	time_t t;
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+	printf("%-8s %-7d %-7d %-7lld\n", ts, e->cpu, e->pid, e->ts/1000);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 data_sz)
+{
+	printf("lost data\n");
+}
+
+static void sig_handler(int sig)
+{
+	env.exiting = true;
+}
+
+static int get_jvmso_path(char *path)
+{
+	char mode[16], line[128], buf[64];
+	size_t seg_start, seg_end, seg_off;
+	FILE *f;
+	int i = 0;
+
+	sprintf(buf, "/proc/%d/maps", env.pid);
+	f = fopen(buf, "r");
+	if (!f)
+		return -1;
+
+	while (fscanf(f, "%zx-%zx %s %zx %*s %*d%[^\n]\n",
+			&seg_start, &seg_end, mode, &seg_off, line) == 5) {
+		i = 0;
+		while (isblank(line[i]))
+			i++;
+		if (strstr(line + i, "libjvm.so")) {
+			break;
+		}
+	}
+
+	strcpy(path, line + i);
+	fclose(f);
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	char binary_path[BINARY_PATH_SIZE] = {0};
+	struct javagc_bpf *skel = NULL;
+	int err;
+	struct perf_buffer *pb = NULL;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	/*
+	* libbpf will auto load the so if it in /usr/lib64 /usr/lib etc,
+	* but the jvmso not there.
+	*/
+	err = get_jvmso_path(binary_path);
+	if (err)
+		return err;
+
+	libbpf_set_strict_mode(LIBBPF_STRICT_ALL);
+	libbpf_set_print(libbpf_print_fn);
+
+	skel = javagc_bpf__open();
+	if (!skel) {
+		fprintf(stderr, "Failed to open BPF skeleton\n");
+		return 1;
+	}
+	skel->bss->time = env.time * 1000;
+
+	err = javagc_bpf__load(skel);
+	if (err) {
+		fprintf(stderr, "Failed to load and verify BPF skeleton\n");
+		goto cleanup;
+	}
+
+	skel->links.handle_gc_start = bpf_program__attach_usdt(skel->progs.handle_gc_start, env.pid,
+									binary_path, "hotspot", "mem__pool__gc__begin", NULL);
+	if (!skel->links.handle_gc_start) {
+		err = errno;
+		fprintf(stderr, "attach usdt mem__pool__gc__begin failed:%s\n",
+				strerror(err));
+		goto cleanup;
+	}
+
+	skel->links.handle_gc_end = bpf_program__attach_usdt(skel->progs.handle_gc_end, env.pid,
+								binary_path, "hotspot", "mem__pool__gc__end", NULL);
+	if (!skel->links.handle_gc_end) {
+		err = errno;
+		fprintf(stderr, "attach usdt mem__pool__gc__end failed:%s\n",
+				strerror(err));
+		goto cleanup;
+	}
+
+	skel->links.handle_gc_start = bpf_program__attach_usdt(skel->progs.handle_gc_start, env.pid,
```

Comment:
`skel->links.handle_gc_start` is assigned twice, this would cause resource leek.

You can add two dummy BPF programs, so the generated skeleton will have two links for attach.

Reply:
You mean add handle_gc_start、handle_gc_end；handle_pool_gc_end、handle_pool_gc_start in javagc.bpf.c

Reply:
Fix it